rfc/change-asset-findings-relationships.md

```diff
@@ -0,0 +1,236 @@
+The data required for the check can be extracted directly from the actual object.
+Uniqueness checks are required for both `Finding` and `AssetFinding` models.
+
+To provide statistical data and aggregation capabilities between other models, _`Finding` should be expanded with the `summary` property_.
+Together, these approaches address the API consistency and data duplication issues.
+
+### Analysis
+
+Take the number of packages that can be discovered on a single production container image as a reference.
+In a real-world scenario, a lot of security findings can be discovered on a given asset.
+Additionally, assets and findings are versioned models where every new version is treated as a completely new object.
+Similarly, asset scans are also versioned in terms of schedule, i.e. new asset scan will be created every week.
+
+Due to the current nature of these models (e.g. creating a new `Finding` or `Asset` for each version, or `AssetScan` on schedule), the size of the tables can grow rapidly.
+The reason to have `AssetFinding` and `AssetScanFinding` as separate tables relates to time and space complexity.
+This is also why a unified table between all three models is not viable.
+
+**Example A**
+To keep track of `#assets = 100`, `#assetScans = 100`, and `#findings = 100` with no versioning changes and no additional scans requires:
+```
+R(asset, assetScan, finding) = #asset * #assetScan * #finding = 100^3 items - Unified table is larger than both other tables combined
+R(asset, finding)            = #asset * #finding              = 100^2 items - AssetFinding table
+R(assetScan, finding)        = #assetScan * #finding          = 100^2 items - AssetScanFinding table
+```
+
+**Example B**
+Unlike the previous example, this takes into account the versioning and scheduling which is a better metric for real-world usage.
+To keep track of `#assets = 100`, `#assetScans = 100`, and `#findings = 100` with _10 new versions_ and _10 new assets for each scan_ requires:
+```
+R(asset, assetScan, finding, version, schedule) = 10^5 * 100^2 - Combined relationship table grows too quickly compared to other tables
+R(asset, finding, version)                      = 10^2 * 100^2 - AssetFinding table depends on assets and findings, i.e. versioning; also transiently depends on scans, but this is omitted for simplicity.
+R(assetScan, finding, version, schedule)        = 10^2 * 100^2 - Scans can be scheduled to occur more often than the versioning changes on other models
+```
+
+Therefore, it makes sense to keep track of associations in separate tables between these models to reduce size and improve efficiency.
+Otherwise, sharding or more complex approaches must be used which are not viable options yet.
+
+### Non-goals
+
+This RFC does not intend to propose changes regarding the relationship between findings and asset scans.
+It is assumed that the asset scan in a given finding represents the latest scan that discovered it.
```

Comment:
what about the `foundOn` timestamp? will we lose the timestamp when a certain finding was first found?

Reply:
just saw that we only keep the last timestamp below, maybe worth adding a new field?

Reply:
The property pair `(foundOn, foundBy)` in this case represents when the finding was last detected because we do not keep the historical data at all. 

Maybe adding an extra field to the summary for `totalDiscovered`  or something similar to indicate how many scans rediscovered the same finding (initial value would be 1). The `(foundOn, foundBy)` can then be used only for the first scan that discovered it. WDYT?

Reply:
> what about the foundOn timestamp? will we lose the timestamp when a certain finding was first found?

Yes, it will be replaced. The property pair `(foundOn, foundBy)` in this case represents when the finding was last detected because we do not keep the historical data. The assumption is that a newer scan might update the data for this finding that does not uniquely describe the finding (e.g. updated description). Can also be switched to denote the first scan. Maybe `refreshedOn` could be used to indicate rediscovery/update by a scan while the other fields can be preserved.

In addition, maybe adding an extra field to the summary for `totalDiscovered`  or something similar to indicate how many scans rediscovered the same finding (initial value would be 1). WDYT?

Reply:
my comment was more on the user perspective where it might be helpful to check how old a finding really is as well as check if it was just found in the last scan.

Reply:
makes sense, thanks for addressing this, will update

Reply:
the new changes should provide a bit more clarity. let me know if this addresses @paralta